File: mcp-server/src/claude_trade/tools/analytics.py
"""成績分析ツール"""

from sqlalchemy import select, func

from claude_trade.db import async_session
from claude_trade.models import DailyPerformance, Order


async def query_performance(
    start_date: str | None = None,
    end_date: str | None = None,
) -> dict:
    """日次成績を集計する

    Args:
        start_date: 開始日 (YYYY-MM-DD, 省略で全期間)
        end_date: 終了日 (YYYY-MM-DD, 省略で全期間)
    """
    async with async_session() as session:
        stmt = select(DailyPerformance).order_by(DailyPerformance.date.asc())
        if start_date:
            stmt = stmt.where(DailyPerformance.date >= start_date)
        if end_date:
            stmt = stmt.where(DailyPerformance.date <= end_date)

        result = await session.execute(stmt)
        days = result.scalars().all()

    if not days:
        return {"message": "No performance data found", "days": []}

    total_pnl = sum(d.pnl for d in days)
    total_trades = sum(d.trades_count for d in days)
    total_winners = sum(d.winners for d in days)
    total_losers = sum(d.losers for d in days)
    win_rate = total_winners / total_trades if total_trades > 0 else 0.0
    max_dd = max((d.max_drawdown_pct for d in days), default=0.0)

    pnl_pcts = [d.pnl_pct for d in days]
    avg_daily_return = sum(pnl_pcts) / len(pnl_pcts) if pnl_pcts else 0.0

    # Simplified Sharpe (annualized, assuming 252 trading days)
    if len(pnl_pcts) > 1:
        import statistics
        std_dev = statistics.stdev(pnl_pcts)
        sharpe = (avg_daily_return / std_dev * (252 ** 0.5)) if std_dev > 0 else 0.0
    else:
        sharpe = 0.0

    return {
        "summary": {
            "total_pnl": round(total_pnl, 2),
            "total_trades": total_trades,
            "win_rate": round(win_rate, 4),
            "max_drawdown_pct": round(max_dd, 4),
            "avg_daily_return_pct": round(avg_daily_return, 4),
            "sharpe_ratio": round(sharpe, 2),
            "trading_days": len(days),
        },
        "daily": [
            {
                "date": d.date,
                "pnl": d.pnl,
                "pnl_pct": d.pnl_pct,
                "trades": d.trades_count,
            }
            for d in days[-30:]  # 直近30日
        ],
    }


async def get_trade_stats(symbol: str | None = None) -> dict:
    """注文統計を取得する

    Args:
        symbol: 特定銘柄に絞る場合のティッカー
    """
    async with async_session() as session:
        stmt = select(Order).where(Order.status == "filled")
        if symbol:
            stmt = stmt.where(Order.symbol == symbol.upper())

        result = await session.execute(stmt)
        orders = result.scalars().all()

    if not orders:
        return {"message": "No filled orders found"}

    buys = [o for o in orders if o.side == "BUY"]
    sells = [o for o in orders if o.side == "SELL"]

    return {
        "total_orders": len(orders),
        "buys": len(buys),
        "sells": len(sells),
        "symbols_traded": list(set(o.symbol for o in orders)),
    }

File: src/mcp-server/tools/analytics.ts
/** 成績分析ツール */

import { eq, gte, lte, asc, and } from "drizzle-orm";
import { db } from "../../db/client.js";
import { dailyPerformance, orders } from "../../db/schema.js";

export async function queryPerformance(
  startDate: string | null = null,
  endDate: string | null = null,
): Promise<Record<string, unknown>> {
  const conditions = [];
  if (startDate) conditions.push(gte(dailyPerformance.date, startDate));
  if (endDate) conditions.push(lte(dailyPerformance.date, endDate));

  const rows = await db
    .select()
    .from(dailyPerformance)
    .where(conditions.length > 0 ? and(...conditions) : undefined)
    .orderBy(asc(dailyPerformance.date));

  if (rows.length === 0) {
    return { message: "No performance data found", days: [] };
  }

  const totalPnl = rows.reduce((sum, d) => sum + d.pnl, 0);
  const totalTrades = rows.reduce((sum, d) => sum + (d.tradesCount ?? 0), 0);
  const totalWinners = rows.reduce((sum, d) => sum + (d.winners ?? 0), 0);
  const totalLosers = rows.reduce((sum, d) => sum + (d.losers ?? 0), 0);
  const winRate = totalTrades > 0 ? totalWinners / totalTrades : 0;
  const maxDd = Math.max(...rows.map((d) => d.maxDrawdownPct ?? 0));

  const pnlPcts = rows.map((d) => d.pnlPct);
  const avgDailyReturn =
    pnlPcts.length > 0
      ? pnlPcts.reduce((a, b) => a + b, 0) / pnlPcts.length
      : 0;

  let sharpe = 0;
  if (pnlPcts.length > 1) {
    const mean = avgDailyReturn;
    const variance =
      pnlPcts.reduce((sum, x) => sum + (x - mean) ** 2, 0) /
      (pnlPcts.length - 1);
    const stdDev = Math.sqrt(variance);
    sharpe = stdDev > 0 ? (mean / stdDev) * Math.sqrt(252) : 0;
  }

  return {
    summary: {
      total_pnl: Math.round(totalPnl * 100) / 100,
      total_trades: totalTrades,
      win_rate: Math.round(winRate * 10000) / 10000,
      max_drawdown_pct: Math.round(maxDd * 10000) / 10000,
      avg_daily_return_pct: Math.round(avgDailyReturn * 10000) / 10000,
      sharpe_ratio: Math.round(sharpe * 100) / 100,
      trading_days: rows.length,
    },
    daily: rows.slice(-30).map((d) => ({
      date: d.date,
      pnl: d.pnl,
      pnl_pct: d.pnlPct,
      trades: d.tradesCount,
    })),
  };
}

export async function getTradeStats(
  symbol: string | null = null,
): Promise<Record<string, unknown>> {
  const conditions = [eq(orders.status, "filled")];
  if (symbol) conditions.push(eq(orders.symbol, symbol.toUpperCase()));

  const rows = await db
    .select()
    .from(orders)
    .where(and(...conditions));

  if (rows.length === 0) {
    return { message: "No filled orders found" };
  }

  const buys = rows.filter((o) => o.side === "BUY");
  const sells = rows.filter((o) => o.side === "SELL");

  return {
    total_orders: rows.length,
    buys: buys.length,
    sells: sells.length,
    symbols_traded: [...new Set(rows.map((o) => o.symbol))],
  };
}
